software/octopi/logging.py

```diff
@@ -0,0 +1,141 @@
+    """
+    if name is None:
+        logger = py_logging.getLogger(_octopi_root_logger_name)
+    else:
+        logger = py_logging.getLogger(_octopi_root_logger_name).getChild(name)
+
+    return logger
+
+
+def set_log_level(level):
+    """
+    All octopi-research code should use this set_log_level method, and the corresponding octopi.logging.get_logger,
+    to control octopi-research-only logging.
+
+    This does not modify the log level of loggers outside the octopi logger hierarchy! If global logging control
+    is needed the normal logging package tools can be used instead.
+    """
+    octopi_root_logger = get_logger()
+    octopi_root_logger.setLevel(level)
+
+    # There's no `getAllChildren` method on the logger or its manager, so we just grab the manager
+    # for our root logger and then check all other loggers to see if they start with our root logger prefix
+    # to find all the octopi specific logger.
+    for (name, logger) in octopi_root_logger.manager.loggerDict.items():
+        # The logging module uses the PlaceHolder object for nodes in the hierarchy that
+        # have children, but no associated loggers.  EG if we create a logger at
+        # octopi.control.gui_hcs but not at octopi.control, then the logger for octopi.control
+        # exists but is a PlaceHolder (until someone explicitly requests it).
+        if name.startswith(_octopi_root_logger_name) and isinstance(logger, py_logging.Logger):
+            logger.setLevel(level)
+
+
+def register_crash_handler(handler, call_existing_too=True):
+    """
+    We want to make sure any uncaught exceptions are logged, so we have this mechanism for putting a hook into
+    the python system that does custom logging when an exception bubbles all the way to the top.
+
+    NOTE: We do our best below, but it is a really bad idea for your handler to raise an exception.
+    """
+    # The sys.excepthook docs are a good entry point for all of this
+    # (here: https://docs.python.org/3/library/sys.html#sys.excepthook), but essentially there are 3 different ways
+    # threads of execution can blow up.  We want to catch and log all 3 of them.
+    old_excepthook = sys.excepthook
+    old_thread_excepthook = threading.excepthook
+    # The unraisable hook doesn't have the same signature as the excepthooks, but we can sort of shoehorn the arguments
+    # into the same signature.  Also, this is an extremely rare (I'm not sure I've ever seen it?) failure mode, so
+    # it should be okay.
+    old_unraisable_hook = sys.unraisablehook
+
+    logger = get_logger()
+
+    def new_excepthook(exception_type: Type[BaseException], value: BaseException, tb: TracebackType):
+        try:
+            handler(exception_type, value, tb)
+        except BaseException as e:
+            logger.critical("Custom excepthook handler raised exception", e)
+        if call_existing_too:
+            old_excepthook(exception_type, value, tb)
```

Comment:
I should rename the tb here and in `new_excepthook` to `tb_type`.  The other two hooks actually get tracebacks (so the `type(tb)` is needed).